subcommand/acl-controller/command.go

```diff
@@ -137,30 +133,64 @@ func (c *Command) run() error {
 		return err
 	}
 
+	ctrl := c.setupController(ecsClient, consulClient, clusterArn)
+
+	for {
+		select {
+		case <-time.After(controller.DefaultPollingInterval):
+			err := ctrl.Reconcile()
+			if err != nil {
+				c.log.Error("error during reconcile", "err", err)
+			}
+		case watcherState := <-c.watcherCh:
+			// TODO: Add unit tests for server watch. Depends on https://github.com/hashicorp/consul-ecs/pull/149
+			c.log.Info("Switching to Consul server", "address", watcherState.Address.String())
+			consulClient, err = c.setupConsulAPIClient(watcherState)
+			if err != nil {
+				c.log.Error("error re-configuring consul client %s", err.Error())
+			} else {
+				ctrl = c.setupController(ecsClient, consulClient, clusterArn)
+			}
+		case <-c.ctx.Done():
+			return nil
+		}
```

Comment:
I'm wondering if we can keep the `Controller.Run` method. I feel like doing so reduces the size of this large-ish `run()` function in the subcommand, and moves this loop functionality into the Controller where the loop logic is more easily testable in isolation.

This is probably a case where it would be nicer to have TaskStateLister call `watcher.State()` (compared to using Subscribe) in order to simply rebuild the api client on-demand. Then the Controller type itself doesn't have to be aware of the Watcher at all. Only the ResourceLister implementations do.

Potentially, we could change TaskStateLister to look like:

```go
// TaskStateLister is an implementation of ResourceLister.
type TaskStateLister struct {
	MakeClientFn func() (*api.Client, error)
	...
}
```

And then when we construct the TaskStateLister, we can do:

```go
	taskStateLister := &controller.TaskStateLister{
		ECSClient:  ecsClient,
		MakeClientFn: c.setupConsulAPIClient,
		ClusterARN: clusterARN,
		Partition:  c.config.Controller.Partition,
		Log:        c.log,
	}
```

And that would need a small change to `setupConsulAPIClient()`:

```go
func (c *Command) setupConsulAPIClient() (*api.Client, error) {
	state, err := c.watcher.State()
	if err != nil {
		return nil, err
	}
	
	// <create and return the client>
}
```

Then TaskStateLister can simply call `MakeClientFn` on-demand, whenever it needs a client. And similarly the the TaskState type could switch to using `MakeClientFn`.

Just one idea. What do you think?